### PR TITLE
Add extra database validations to CreateDatabase

### PR DESCRIPTION
**Pull request by greedy52**
Related issue #17454

Why
When creating databases through static config, we have a bunch of awesome pre-checks to validate the inputs. It would be nice to have the same checks when creating database through the webapi (or `tctl`)

Changes
- Moved `redis.ParseRedisAddress` from `lib/srv/db/redis/connection.go` its own package `lib/srv/db/redis/connection`
- Moved database validation from `service.Database.CheckAndSetDefaults` to `services.ValidateDatabase` (and one check to `api.DatabaseV3.CheckAndSetDefaults`)
- Added validation to `services/local/databases.go` similar to other validations in `services/local`

### CODE REVIEW

File: lib/services/local/databases.go [hunk 1]
```diff
@@ -74,7 +74,7 @@ func (s *DatabaseService) GetDatabase(ctx context.Context, name string) (types.D
 
 // CreateDatabase creates a new database resource.
 func (s *DatabaseService) CreateDatabase(ctx context.Context, database types.Database) error {
-	if err := database.CheckAndSetDefaults(); err != nil {
+	if err := services.ValidateDatabase(database); err != nil {
```
Comment (greedy52) [inline lib/services/local/databases.go:77]: this is the only real change, rest are moving things around.

File: api/types/database.go [hunk 2]
```diff
@@ -605,6 +605,14 @@ func (d *DatabaseV3) CheckAndSetDefaults() error {
...
 		}
 	}
+
+	// Validate Cloud SQL specific configuration.
+	switch {
+	case d.Spec.GCP.ProjectID != "" && d.Spec.GCP.InstanceID == "":
+		return trace.BadParameter("missing Cloud SQL instance ID for database %q", d.GetName())
+	case d.Spec.GCP.ProjectID == "" && d.Spec.GCP.InstanceID != "":
+		return trace.BadParameter("missing Cloud SQL project ID for database %q", d.GetName())
+	}
```
Comment (marcoandredinis) [inline api/types/database.go:615]: Maybe create a separate function for GCP checks?

File: lib/service/cfg.go [hunk 3]
```diff
@@ -947,75 +942,17 @@ func (d *DatabaseAD) CheckAndSetDefaults(name string) error {
...
-			d.Name, d.URI, err)
-	}
 
-	if len(d.TLS.CACert) != 0 {
-		if _, err := tlsca.ParseCertificatePEM(d.TLS.CACert); err != nil {
-			return trace.BadParameter("provided database %q CA doesn't appear to be a valid x509 certificate: %v",
-				d.Name, err)
-		}
-	}
+	// Check and set defaults for TLS mode.
```
Comment (marcoandredinis) [inline lib/service/cfg.go:951]: :nail_care: 
Superfluous comment

**Review by marcoandredinis — APPROVED** [review]

File: lib/service/cfg.go [hunk 4]
```diff
@@ -1028,7 +965,68 @@ func (d *Database) CheckAndSetDefaults() error {
...
+	// Do a test run with extra validations.
+	db, err := d.ToDatabase()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(services.ValidateDatabase(db))
+}
+
+// ToDatabase converts Database to types.Database.
+func (d Database) ToDatabase() (types.Database, error) {
```
Comment (smallinsky) [inline lib/service/cfg.go:977]:
```suggestion
func (d *Database) ToDatabase() (types.Database, error) {
```

**Review by smallinsky — APPROVED** [review]

**Comment by greedy52** [comment]
Thanks for the quick review 😄

**Pull request merged**